From f63f6975ccdeeb979a6f32c71a5e17a7d76c5544 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Wed, 01 Dec 2021 18:05:02 +0200
Subject: [PATCH] [SW-36216] gaudi2: sec: reroute lbw tpc traffic

As dcore0 rtr0 lbw if is a shared interface for pcie, tpc0 and tpc1
it is required to re-route the traffic for tpc0 and tpc1 such that
only pcie traffic is allowed through dcore0 rtr0. This flexibility
is present for lbw ifs. The routing config block is connected only
in dcore0 rtr2 block in hardware, hence we program the dcore0 rtr0
range register whilst configuration of trafiic re-reroute is done
in dcore0 rtr2.

Change-Id: I3eecfe5de727e10775df805e3dcfa4793f641f95
---

diff --git a/arch/gaudi2/common/security.c b/arch/gaudi2/common/security.c
index ca953e1..86ea237 100644
--- a/arch/gaudi2/common/security.c
+++ b/arch/gaudi2/common/security.c
@@ -26,6 +26,8 @@
 #include <dcore0_rtr0_mstr_if_rr_shrd_hbw_regs.h>
 #include <jt_mstr_if_rr_shrd_hbw_regs.h>
 #include <sft0_hbw_rtr_if0_mstr_if_rr_shrd_hbw_regs.h>
+#include <dcore0_rtr2_masks.h>
+#include <dcore0_rtr2_regs.h>
 
 /*
  * instances_num : Number of instances of RR to program.
@@ -833,6 +835,25 @@
 	hl_secure_dyn_rr_cfg(cfg->rr, cfg->min, cfg->max);
 }
 
+static void preboot_lbw_rtr_preconfig(void)
+{
+	/*
+	 * As dcore0 rtr0 lbw if is a shared interface for pcie, tpc0 and tpc1
+	 * it is required to re-route the traffic for tpc0 and tpc1 such that
+	 * only pcie traffic is allowed through dcore0 rtr0. This flexibility
+	 * is present for lbw ifs. The routing config block is connected only
+	 * in dcore0 rtr2 block in hardware, hence we program the dcore0 rtr0
+	 * range register whilst configuration of trafiic re-reroute is done
+	 * in dcore0 rtr2.
+	 */
+	u32 reg = regmap_readl(mmDCORE0_RTR2_TIF_LBW_USR_RTR);
+
+	reg |= BIT(DCORE0_RTR2_TIF_LBW_USR_RTR_TPC0_RTR_SHIFT);
+	reg |= BIT(DCORE0_RTR2_TIF_LBW_USR_RTR_TPC1_RTR_SHIFT);
+
+	regmap_writel(reg, mmDCORE0_RTR2_TIF_LBW_USR_RTR);
+}
+
 void ppboot_security_init(void)
 {
 	preboot_secure_arc();
@@ -851,6 +872,7 @@
 	if (!preboot_security_is_enabled())
 		return;
 
+	preboot_lbw_rtr_preconfig();
 	hl_dev_prot_bits_init();
 	preboot_secure_lbw();
 	preboot_secure_hbw();
